### PR TITLE
servicemgr: update mount files

### PR DESCRIPTION
**Pull request by valentindavid**
When updating service files for the snaps, this also verifies the mount unit of the snap is updated. This however does not update mounts generated from `snapctl mount`.

Note that there is a lot or renaming. So to ease the review I have split it in commit with separate concerns.

### CODE REVIEW

File: systemd/systemd.go [hunk 1]
```diff
@@ -326,6 +326,14 @@ const (
 	EmulationModeBackend
 )
 
+type MountUpdateStatus int
+
+const (
+	MountClean MountUpdateStatus = iota
```
Comment (mardy) [inline systemd/systemd.go:332]: These values need some explanatory comments, as it's not clear to me what "MountClean" means.

Comment (mardy) [reply systemd/systemd.go:332]: If I understand correctly the code, MountClean could be renamed to MountUnchanged, and and the comment should probably say that this does not guarantee that the mount exists and that it literally means that the preexisting situation was unchanged.

Maybe it could say that if MountClean is returned with a `nil` error then it means that the mount exists.

File: systemd/systemd.go [hunk 2]
```diff
@@ -1457,23 +1484,24 @@ func (s *systemd) AddMountUnitFileWithOptions(unitOptions *MountUnitOptions) (st
...
-	if err := s.EnableNoReload(units); err != nil {
-		return "", err
-	}
-	if err := s.Start(units); err != nil {
-		return "", err
+		units := []string{mountUnitName}
+		if err := s.EnableNoReload(units); err != nil {
+			return "", err
+		}
+		if err := s.ReloadOrRestart(units); err != nil {
```
Comment (mardy) [inline systemd/systemd.go:1502]: I think that this change (from `Start` to `ReloadOrRestart` needs to be explained better, either as a code comment or in the commit message if it's long).

File: systemd/systemd.go [hunk 3]
```diff
@@ -1457,23 +1484,24 @@ func (s *systemd) AddMountUnitFileWithOptions(unitOptions *MountUnitOptions) (st
 	daemonReloadLock.Lock()
 	defer daemonReloadLock.Unlock()
 
-	mountUnitName, err := writeMountUnitFile(unitOptions)
+	mountUnitName, modified, err := writeMountUnitFile(unitOptions)
 	if err != nil {
 		return "", err
 	}
+	if modified != MountClean {
```
Comment (mardy) [inline systemd/systemd.go:1491]: I would expect this change to have a major effect on our unit tests, but it looks like the only reason they had to be changed was the switch from `Start` to `ReloadOrRestart`, which makes me wonder if the test coverage here is good enough. But in any case, it would be nice to have a unit test that exercises the code path when `modified` is equal to `MountClean`.

File: wrappers/core18.go [hunk 4]
```diff
@@ -64,8 +64,8 @@ func writeSnapdToolingMountUnit(sysd systemd.Systemd, prefix string, opts *AddSn
 
 	// TODO: the following comment is wrong, we don't need RequiredBy=snapd here?
 
-	// Not using AddMountUnitFile() because we need
-	// "RequiredBy=snapd.service"
+	// Not using EnsureMountUnitFile() because we need
+	// "WantedBy=snapd.service"
```
Comment (mardy) [inline wrappers/core18.go:68]: I think this change should go to a separate commit, with some explanation.

File: overlord/servicestate/servicemgr.go [hunk 5]
```diff
@@ -108,6 +108,31 @@ func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
...
+			return err
+		}
+		squashfsPath := dirs.StripRootDir(info.MountFile())
+		whereDir := dirs.StripRootDir(info.MountDir())
+		var sysd systemd.Systemd
+		if snapdenv.Preseeding() {
+			sysd = systemd.NewEmulationMode(dirs.GlobalRootDir)
+		} else {
+			sysd = systemd.New(systemd.SystemMode, nil)
+		}
```
Comment (mardy) [inline overlord/servicestate/servicemgr.go:123]: This could be taken out of the loop (maybe behind a `len(allStates) > 0`, if it's expensive -- I don't know)

File: tests/main/snapd-update-services/task.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
+summary: systemd units get regnerated when snapd is refreshed
```
Comment (mardy) [inline tests/main/snapd-update-services/task.yaml:1]: regenerated

File: tests/main/snapd-update-services/task.yaml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  - ubuntu-18.04-*
+  - ubuntu-2*
+
+prepare: |
+  # Install a snap with a service running
+  service_snap="$("${TESTSTOOLS}/snaps-state" pack-local test-snapd-service)"
+  snap install --dangerous "${service_snap}"
+  systemctl is-active snap.test-snapd-service.test-snapd-service.service
+
+  #. "$TESTSLIB"/prepare.sh
```
Comment (mardy) [inline tests/main/snapd-update-services/task.yaml:17]: Remove this, if it's not needed

File: tests/main/snapd-update-services/task.yaml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  unpackdir=/tmp/snapd-update-services
+  snap download --edge snapd --basename=snapd_edge
+  unsquashfs -no-progress -d "${unpackdir}" snapd_edge.snap
+  #rm -f "$UNPACK_DIR"/etc/apparmor.d/*
```
Comment (mardy) [inline tests/main/snapd-update-services/task.yaml:21]: Also this one

File: tests/main/snapd-update-services/task.yaml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  dpkg-deb -x "${GOHOME}"/snapd_*.deb "${unpackdir}"
+  cp /usr/lib/snapd/info "${unpackdir}"/usr/lib/snapd
```
Comment (mardy) [inline tests/main/snapd-update-services/task.yaml:23]: Please add a comment explaining what we are doing

File: tests/main/snapd-update-services/task.yaml [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    systemctl reload-or-restart "${unit}"
+  done
+
+  # We simulate Options=dev to Options=nodev update
+  mount_units=()
+  for f in /etc/systemd/system/snap-*.mount; do
+    unit="$(basename "${f}")"
+    case "${unit}" in
+      snap-snapd-*.mount)
+        # This version version will get disabled and not updated
```
Comment (mardy) [inline tests/main/snapd-update-services/task.yaml:51]: duplicate "version"

**Review by mardy — COMMENTED** [review]
Reviewed comit by commit, hopefully the comments are not lost :-)

LGTM, but I think it needs some clarifications here and there.

File: systemd/systemd.go (hunk 1 above)
Comment (valentindavid) [reply systemd/systemd.go:332]: I think `MountUnchanged` is better, yes.

The value returned when the error is not `nil` is relevant. It is the same as the mount unit name being an empty string.

File: tests/main/snapd-update-services/task.yaml (hunk 9 above)
Comment (valentindavid) [reply tests/main/snapd-update-services/task.yaml:23]: I copied some code from `prepare.sh` and forgot to clean it up. That line is not useful.

File: overlord/servicestate/servicemgr.go (hunk 5 above)
Comment (valentindavid) [reply overlord/servicestate/servicemgr.go:123]: Good point. It might be uselessly expensive on the garbage collection. I will move it.

File: wrappers/core18.go (hunk 4 above)
Comment (valentindavid) [reply wrappers/core18.go:68]: That should not be here. I do not even understand the comment.

File: systemd/systemd.go (hunk 3 above)
Comment (valentindavid) [reply systemd/systemd.go:1491]: In the case the unit was created, the service will not be started and in this case `ReloadOrRestart` will have the same behavior as `Start`. So this is why I did not split the case `MountUpdated` and `MountCreated`.

That the case `MountClean` is not tested in unit test was a clear oversight however. I will add it.

File: systemd/systemd.go (hunk 2 above)
Comment (valentindavid) [reply systemd/systemd.go:1506]: I added a comment.

File: systemd/systemd.go (hunk 3 above)
Comment (valentindavid) [reply systemd/systemd.go]: The test is `TestEnureMountUnitUnchanged`.

File: wrappers/core18.go (hunk 4 above)
Comment (valentindavid) [reply wrappers/core18.go]: I have removed this change.

File: overlord/servicestate/servicemgr.go (hunk 5 above)
Comment (valentindavid) [reply overlord/servicestate/servicemgr.go]: Done.

File: tests/main/snapd-update-services/task.yaml (hunk 6 above)
Comment (valentindavid) [reply tests/main/snapd-update-services/task.yaml]: Done.

File: tests/main/snapd-update-services/task.yaml (hunk 7 above)
Comment (valentindavid) [reply tests/main/snapd-update-services/task.yaml]: Done.

File: tests/main/snapd-update-services/task.yaml (hunk 8 above)
Comment (valentindavid) [reply tests/main/snapd-update-services/task.yaml]: Done.

File: tests/main/snapd-update-services/task.yaml (hunk 9 above)
Comment (valentindavid) [reply tests/main/snapd-update-services/task.yaml]: Removed.

File: tests/main/snapd-update-services/task.yaml (hunk 10 above)
Comment (valentindavid) [reply tests/main/snapd-update-services/task.yaml]: Done.

**Review by mardy — COMMENTED** [review]
There's a small error in the tests, but already LGTM! Thanks for the updates!

**Review by mardy — APPROVED** [review]
Thanks, LGTM!

File: overlord/servicestate/servicemgr.go [hunk 11]
```diff
@@ -117,6 +117,34 @@ func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
 		return nil
 	}
 
+	var sysd systemd.Systemd
```
Comment (pedronis) [inline overlord/servicestate/servicemgr.go:120]: no strong opinion, but conceptually this belongs to snapstate (as it is the manager that creates the units to start with) and not here

File: systemd/systemd.go [hunk 12]
```diff
@@ -326,6 +326,14 @@ const (
...
 )
 
+type MountUpdateStatus int
+
+const (
+	MountUnchanged MountUpdateStatus = iota
+	MountUpdated
+	MountCreated
+)
+
```
Comment (pedronis) [inline systemd/systemd.go:336]: not sure this needs to exported, they seem to be used only internally

File: systemd/systemd.go [hunk 13]
```diff
@@ -1392,22 +1400,41 @@ const (
...
+			return "", MountUnchanged, err
+		}
+		outf.Seek(0, os.SEEK_SET)
+		if osutil.StreamsEqual(outf, f) {
+			return filepath.Base(mu), MountUnchanged, nil
+		}
+		if err != nil {
+			return "", MountUnchanged, err
+		}
+		modified = MountUpdated
```
Comment (pedronis) [inline systemd/systemd.go:1434]: the tests don't seem to reach here

File: systemd/systemd.go [hunk 14]
```diff
@@ -1392,22 +1400,41 @@ const (
...
-		return "", fmt.Errorf("cannot generate mount unit: %v", err)
+		return "", MountUnchanged, fmt.Errorf("cannot generate mount unit: %v", err)
+	}
+
+	f, err := os.Open(mu)
+	defer f.Close()
+	if os.IsNotExist(err) {
+		modified = MountCreated
+	} else {
+		if err != nil {
```
Comment (pedronis) [inline systemd/systemd.go:1424]: I wonder if instead of this code we shouldn't move this code to use osutil.EnsureFileState. that one doesn't what kind of changes it made but it could be made to do so, as other functions in its family do that

File: systemd/emulation.go [hunk 15]
```diff
@@ -147,12 +147,23 @@ func (s *emulation) AddMountUnitFile(snapName, revision, what, where, fstype str
 		return "", err
 	}
 
+	if modified == MountUnchanged {
+		return mountUnitName, nil
```
Comment (pedronis) [inline systemd/emulation.go:151]: in emulation I would expect that changing or preexisting units are unlikely but it still a good idea to support them, tests don't seem to reach the MountUnchanged and MountUpdated cases here

File: overlord/servicestate/servicemgr.go [hunk 16]
```diff
@@ -117,6 +117,34 @@ func (m *ServiceManager) ensureSnapServicesUpdated() (err error) {
...
+			info.InstanceName(),
+			info.Revision.String(),
+			squashfsPath,
+			whereDir,
+			"squashfs",
+		)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (pedronis) [inline overlord/servicestate/servicemgr.go:146]: the changes made to servicemgr_test.go don't seem to care if the code is here or removed

**Review by pedronis — COMMENTED** [review]
thank you, some questions/comments about placement, the implementation of the ensure helper vs osutil changes, and tests

File: overlord/servicestate/servicemgr.go (hunk 11 above)
Comment (valentindavid) [reply overlord/servicestate/servicemgr.go:120]: Do you mean to move the content of the loop after into a method in snapstate, but still call it from ServiceManager.ensureSnapServicesUpdated? Otherwise I am a bit confused.

File: overlord/servicestate/servicemgr.go (hunk 16 above)
Comment (valentindavid) [reply overlord/servicestate/servicemgr.go:146]: I have added 3 tests.

File: systemd/emulation.go (hunk 15 above)
Comment (valentindavid) [reply systemd/emulation.go:151]: I have added some tests.

File: systemd/systemd.go (hunk 14 above)
Comment (valentindavid) [reply systemd/systemd.go]: It now uses `osutil.EnsureFileState`.

File: systemd/systemd.go (hunk 12 above)
Comment (valentindavid) [reply systemd/systemd.go]: Done.

File: overlord/servicestate/servicemgr.go (hunk 11 above)
Comment (valentindavid) [reply overlord/servicestate/servicemgr.go:120]: Or did you mean it should be called from `SnapManager.Ensure`?

Comment (pedronis) [reply overlord/servicestate/servicemgr.go:120]: the latter, this part going under SnapManager.Ensure

Comment (valentindavid) [reply overlord/servicestate/servicemgr.go:120]: OK. I have managed to get it to work with the spread test. But now I have to modify all the tests in `snapstate_test.go`.

Comment (pedronis) [reply overlord/servicestate/servicemgr.go:120]: why all the tests? maybe I'm misunderstanding, what we do sometimes in these cases is expose a way in export_test.go to set a flag on the manager to avoid some of the Ensure logic when it's not the test focus. 

see SetBootOkRan for example in devicestate/export_test.go

File: osutil/io.go [hunk 17]
```diff
@@ -98,7 +98,7 @@ func NewAtomicFile(filename string, perm os.FileMode, flags AtomicWriteFlags, ui
 	// suffixes, such as the one we selected below.
 	tmp := filename + "." + randutil.RandomString(12) + "~"
 
-	fd, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC|os.O_EXCL, perm)
+	fd, err := os.OpenFile(tmp, os.O_RDWR|os.O_CREATE|os.O_TRUNC|os.O_EXCL, perm)
```
Comment (pedronis) [inline osutil/io.go:101]: do we still need this change?

File: overlord/servicestate/servicemgr.go (hunk 11 above)
Comment (pedronis) [reply overlord/servicestate/servicemgr.go:120]: any news on this front?

File: systemd/systemd.go [hunk 18]
```diff
@@ -1392,22 +1401,39 @@ const (
...
+		return "", mountUnchanged, fmt.Errorf("cannot generate mount unit: %v", err)
 	}
-	return filepath.Base(mu), outf.Commit()
+
+	if osutil.FileExists(mu) {
+		modified = mountUpdated
+	} else {
+		modified = mountCreated
+	}
+	stateErr := osutil.EnsureFileState(mu, &osutil.FileReferencePlusMode{
```
Comment (pedronis) [inline systemd/systemd.go:1425]: instead of a TempFile this now could use a bytes.Buffer and a osutil.MemoryFileState, no?

**Review by pedronis — APPROVED** [review]
did another pass, some questions/comments

File: overlord/servicestate/servicemgr.go (hunk 11 above)
Comment (valentindavid) [reply overlord/servicestate/servicemgr.go]: Sorry, I forgot to finish it. Now it is pushed.

File: systemd/systemd.go (hunk 18 above)
Comment (valentindavid) [reply systemd/systemd.go:1425]: `template.Template.Execute` expects a file. I did not find anything in the standard go library or in osutil that provides a writable memory file. Is there anything I can use?

File: osutil/io.go (hunk 17 above)
Comment (valentindavid) [reply osutil/io.go]: I have dropped this commit.

File: systemd/systemd.go (hunk 18 above)
Comment (valentindavid) [reply systemd/systemd.go:1425]: I meant an `io.Writer`.

Comment (pedronis) [reply systemd/systemd.go:1425]: I'm confused by this.  bytes.Buffer is a io.Writer and you can find other examples in the code base using template and bytes.Buffer together, see wrappers/services.go for example

Comment (valentindavid) [reply systemd/systemd.go]: Done.

**Comment by pedronis** [comment]
this needs  a bit more work as it seems some unit tests are failing and some are not mocking systemctl and one gets prompted about allowing real operations

File: overlord/snapstate/snapmgr.go [hunk 19]
```diff
@@ -1067,6 +1068,56 @@ func (m *SnapManager) localInstallCleanup() error {
 	return osutil.UnlinkManyAt(d, filenames)
 }
 
+func (m *SnapManager) ensureMountUpdated() error {
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:1071]: s/Mount/Mounts/ ?

**Review by pedronis — CHANGES_REQUESTED** [review]
I'm still getting prompted to allow systemd reload when running snapstate tests so it still seems to be missing some mocking via systemd.MockSystemctl. Also some comments about how often we should run ensureMountUpdated and how to deal with it in tests. The latter might help also with the first problem is some cases.

This kind of change can help find the first issue:
```
diff --git a/systemd/systemd.go b/systemd/systemd.go
index 072a6ef724..49460653ce 100644
--- a/systemd/systemd.go
+++ b/systemd/systemd.go
@@ -120,6 +120,9 @@ func MockNewSystemd(f func(be Backend, rootDir string, mode InstanceMode, rep Re
 var systemctlCmd = func(args ...string) ([]byte, error) {
        // TODO: including stderr here breaks many things when systemd is in debug
        // output mode, see LP #1885597
+       if len(args) != 1 || args[0] != "--version" {
+               panic(fmt.Sprintf("SYSTEMCTL %v", args))
+       }
        bs, err := exec.Command("systemctl", args...).CombinedOutput()
        if err != nil {
                exitCode, runErr := osutil.ExitCode(err)
```

File: overlord/snapstate/snapmgr.go [hunk 20]
```diff
@@ -1067,6 +1068,56 @@ func (m *SnapManager) localInstallCleanup() error {
...
+	m.state.Lock()
+	defer m.state.Unlock()
+
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply overlord/snapstate/snapmgr.go:1074]: Do you want the function to be in snapstate, but be called from servicestate? It was originally in servicestate, but you said previously:

> no strong opinion, but conceptually this belongs to snapstate (as it is the manager that creates the units to start with) and not here

[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:1074]: no, I asked to move it, this is the right place. What I was trying to say is that this new method should follow more of the patterns of ServiceManager.ensureSnapServicesUpdated (having a flag on the manager, having a way to mock the flag state) to avoid some of the issues we have here atm

[the remark this answers is not in the document]
Comment (valentindavid) [reply overlord/snapstate/snapmgr.go:1086]: Done.

File: overlord/snapstate/snapmgr.go (hunk 19 above)
Comment (valentindavid) [reply overlord/snapstate/snapmgr.go]: Done.

**Comment by valentindavid** [comment]
There are some tests that seem to fail. I have to look at that.

File: overlord/snapstate/snapstate_test.go [hunk 21]
```diff
@@ -119,6 +121,24 @@ func (s *snapmgrBaseTest) logTasks(c *C) {
...
+func (s *snapmgrBaseTest) mockSystemctlCallsUpdateMounts(c *C) (restore func()) {
+	r := systemd.MockSystemctl(func(args ...string) ([]byte, error) {
+		if len(args) == 1 && args[0] == "daemon-reload" {
+			return []byte(""), nil
+		}
+		if len(args) == 3 && args[0] == "--no-reload" && args[1] == "enable" {
+			return []byte(""), nil
+		}
+		if len(args) == 2 && args[0] == "reload-or-restart" {
+			return []byte(""), nil
```
Comment (pedronis) [inline overlord/snapstate/snapstate_test.go:133]: should we also track/count if reload-or-restart is called to distinguish situations in  the new tests below?

**Review by pedronis — APPROVED** [review]
thank you, one comment about tests

File: overlord/snapstate/snapmgr.go [hunk 22]
```diff
@@ -1067,6 +1071,71 @@ func (m *SnapManager) localInstallCleanup() error {
...
+	}
+
+	for _, snapSt := range allStates {
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+		squashfsPath := dirs.StripRootDir(info.MountFile())
+		whereDir := dirs.StripRootDir(info.MountDir())
+		_, err = sysd.EnsureMountUnitFile(
```
Comment (mvo5) [inline overlord/snapstate/snapmgr.go:1122]: (super-nitpick) this could be a single line: `if _, err := sysd.EnsureMountUnitFile(...); err != nil ` (it looks like even then it's not *that* long. But feel free to ignore, it's really a minor nitpick.

File: overlord/snapstate/snapstate_test.go [hunk 23]
```diff
@@ -7314,8 +7336,10 @@ func (s *snapmgrTestSuite) TestStartSnapServicesWithDisabledServicesNowApp(c *C)
 
 	s.runStartSnapServicesWithDisabledServices(c, "hello")
 
+	re := regexp.MustCompile(`.*previously disabled service hello is now an app and not a service\n.*`)
+
 	// check the log for the notice
-	c.Assert(buf.String(), Matches, `.*previously disabled service hello is now an app and not a service\n.*`)
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go:7318]: Out of curiosity, why this change ? It seems the original was just fine? It probably need a regexp flag though, this:
```
c.Assert(buf.String(), Matches, `(?s).*previously disabled service hello is now an app and not a service\n.*`)
```

should work (here and below).

File: overlord/snapstate/snapstate_test.go [hunk 24]
```diff
@@ -8631,3 +8657,191 @@ func (s *snapmgrTestSuite) TestResolveValidationSetsEnforcementErrorWithInvalidS
 	_, _, err := snapstate.ResolveValidationSetsEnforcementError(context.TODO(), s.state, valErr, nil, s.user.ID)
 	c.Assert(err, ErrorMatches, "cannot auto-resolve validation set constraints that require removing snaps: \"snap-a\"")
 }
+
+func (s *snapmgrTestSuite) TestEnsureSnapStateRewriteMounts(c *C) {
+	s.state.Lock()
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go:8662]: I would move the lock down closer to where it is actually needed:
```
@@ -8659,8 +8657,6 @@ func (s *snapmgrTestSuite) TestResolveValidationSetsEnforcementErrorWithInvalidS
 }
 
 func (s *snapmgrTestSuite) TestEnsureSnapStateRewriteMounts(c *C) {
-       s.state.Lock()
-
        restore := snapstate.MockEnsuredMountsUpdated(s.snapmgr, false)
        defer restore()
 
@@ -8678,9 +8674,9 @@ apps:
     command: bin.sh
 `
 
+       s.state.Lock()
        snapstate.Set(s.state, "test-snap", testSnapState)
        snaptest.MockSnapCurrent(c, testYaml, testSnapSideInfo)
-
        s.state.Unlock()
 
        what := fmt.Sprintf("%s/%s_%s.snap", "/var/lib/snapd/snaps", "test-snap", "42")

```

File: overlord/snapstate/snapstate_test.go [hunk 25]
```diff
@@ -8631,3 +8657,191 @@ func (s *snapmgrTestSuite) TestResolveValidationSetsEnforcementErrorWithInvalidS
...
+[Install]
+WantedBy=snapd.mounts.target
+WantedBy=multi-user.target
+`[1:], what)
+
+	c.Assert(mountFile, testutil.FileEquals, expectedContent)
+}
+
+func (s *snapmgrTestSuite) TestEnsureSnapStateRewriteMountsNoChange(c *C) {
+	s.state.Lock()
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go:8737]: Same suggestion here as above.

File: overlord/snapstate/snapstate_test.go [hunk 26]
```diff
@@ -8631,3 +8657,191 @@ func (s *snapmgrTestSuite) TestResolveValidationSetsEnforcementErrorWithInvalidS
...
+	c.Assert(err, IsNil)
+
+	err = s.snapmgr.Ensure()
+	c.Assert(err, IsNil)
+
+	c.Assert(mountFile, testutil.FileEquals, mountContent)
+}
+
+func (s *snapmgrTestSuite) TestEnsureSnapStateRewriteMountsCreated(c *C) {
+	s.state.Lock()
```
Comment (mvo5) [inline overlord/snapstate/snapstate_test.go:8793]: And here :)

File: systemd/emulation.go [hunk 27]
```diff
@@ -136,7 +136,7 @@ func (s *emulation) AddMountUnitFile(snapName, revision, what, where, fstype str
 	// This means that when preseeding in a lxd container, the snap will be
 	// mounted with fuse, but mount unit will use squashfs.
 	mountUnitOptions := append(fsMountOptions(fstype), squashfs.StandardOptions()...)
-	mountUnitName, err := writeMountUnitFile(&MountUnitOptions{
+	mountUnitName, modified, err := ensureMountUnitFile(&MountUnitOptions{
```
Comment (mvo5) [inline systemd/emulation.go:139]: This is nice!

File: systemd/systemd.go [hunk 28]
```diff
@@ -1392,22 +1400,37 @@ const (
...
 
-	if err := parsedMountUnitTemplate.Execute(outf, &u); err != nil {
-		return "", fmt.Errorf("cannot generate mount unit: %v", err)
+	if osutil.FileExists(mu) {
+		modified = mountUpdated
+	} else {
+		modified = mountCreated
+	}
+
+	os.MkdirAll(filepath.Dir(mu), 0755)
```
Comment (mvo5) [inline systemd/systemd.go:1421]: Is here an error check `if err := os.MkidrAll(...` missing?

File: systemd/systemd_test.go [hunk 29]
```diff
@@ -1185,7 +1185,133 @@ WantedBy=multi-user.target
...
+[Install]
+WantedBy=snapd.mounts.target
+WantedBy=multi-user.target
+`[1:], mockSnapPath)
+	err = ioutil.WriteFile(filepath.Join(dirs.SnapServicesDir, "snap-snapname-123.mount"), []byte(content), 0644)
+	c.Assert(err, IsNil)
+
+	mountUnitName, err := NewUnderRoot(rootDir, SystemMode, nil).EnsureMountUnitFile("foo", "42", mockSnapPath, "/snap/snapname/123", "squashfs")
+	c.Assert(err, IsNil)
+	defer os.Remove(mountUnitName)
```
Comment (mvo5) [inline systemd/systemd_test.go:1227]: Is this needed? Given that rootDir is under a c.MkDir() anyway it will get cleaned up already or am I missing something?

File: systemd/systemd_test.go [hunk 30]
```diff
@@ -1185,7 +1185,133 @@ WantedBy=multi-user.target
...
```
Comment (mvo5) [inline systemd/systemd_test.go:1288]: Same question as above

File: systemd/systemd_test.go [hunk 31]
```diff
@@ -1692,6 +1818,100 @@ func (s *SystemdTestSuite) TestPreseedModeAddMountUnit(c *C) {
...
+[Install]
+WantedBy=snapd.mounts.target
+WantedBy=multi-user.target
+`[1:], mockSnapPath)
+	err = ioutil.WriteFile(filepath.Join(dirs.SnapServicesDir, "snap-snapname-123.mount"), []byte(content), 0644)
+	c.Assert(err, IsNil)
+
+	mountUnitName, err := sysd.EnsureMountUnitFile("foo", "42", mockSnapPath, "/snap/snapname/123", "squashfs")
+	c.Assert(err, IsNil)
+	defer os.Remove(mountUnitName)
```
Comment (mvo5) [inline systemd/systemd_test.go:1858]: Same question as above

File: systemd/systemd_test.go [hunk 32]
```diff
@@ -1692,6 +1818,100 @@ func (s *SystemdTestSuite) TestPreseedModeAddMountUnit(c *C) {
...
```
Comment (mvo5) [inline systemd/systemd_test.go:1904]: and again :)

**Review by mvo5 — COMMENTED** [review]
Thanks, this looks super nice. Some small suggestions and ideas and questions but nothing looks like a blocker

File: overlord/snapstate/snapstate_test.go (hunk 23 above)
Comment (valentindavid) [reply overlord/snapstate/snapstate_test.go:7318]: Must be merge conflict wrongly solved.

File: systemd/systemd.go (hunk 28 above)
Comment (valentindavid) [reply systemd/systemd.go:1421]: It is because it does not matter much. It is quite common to not care about failure from a recursive mkdir. But looking the rest of snapd, this project always checks. I can add the check.

File: overlord/snapstate/snapstate_test.go (hunk 23 above)
Comment (valentindavid) [reply overlord/snapstate/snapstate_test.go]: Ah, right, now I remember. That works.

**Review by mvo5 — APPROVED** [review]
Thanks, this looks great now

File: overlord/snapstate/snapmgr.go [hunk 33]
```diff
@@ -1067,6 +1071,64 @@ func (m *SnapManager) localInstallCleanup() error {
 	return osutil.UnlinkManyAt(d, filenames)
 }
 
+func MockEnsuredMountsUpdated(m *SnapManager, ensured bool) (restore func()) {
+	osutil.MustBeTestBinary("ensured snap mounts can only be mocked from tests")
+	old := m.ensuredMountsUpdated
```
Comment (Meulengracht) [inline overlord/snapstate/snapmgr.go:1076]:
```suggestion
	old := m.ensuredMountsUpdated
```

**Review by Meulengracht — COMMENTED** [review]
This looks really nice, some comments on the `ensureMountsUpdated`

File: overlord/snapstate/snapmgr.go [hunk 34]
```diff
@@ -1067,6 +1071,64 @@ func (m *SnapManager) localInstallCleanup() error {
...
+	m.ensuredMountsUpdated = ensured
+	return func() {
+		m.ensuredMountsUpdated = old
+	}
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:1080]: no, it is used by other package afair? also we can't use testutil from production code, so unless we are in export_test.go no testutil.Backup for us. Life is hard.

[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/snapstate/snapmgr.go:1080]: Move it into snapstatetest package and use testutil.Backup instead?

[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:1080]: can't do that either because that package can't touch the unexported flag. The code is fine as is really. testutil.Backup is convenient but not worth making the code a pretzel over it.

[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/snapstate/snapmgr.go:1080]: Yea wasn't so worried about the testutil.Backup, just wanted to avoid Test code in production code

File: overlord/snapstate/snapmgr.go [hunk 35]
```diff
@@ -1067,6 +1071,64 @@ func (m *SnapManager) localInstallCleanup() error {
...
+		info, err := snapSt.CurrentInfo()
+		if err != nil {
+			return err
+		}
+		squashfsPath := dirs.StripRootDir(info.MountFile())
+		whereDir := dirs.StripRootDir(info.MountDir())
+		if _, err = sysd.EnsureMountUnitFile(info.InstanceName(), info.Revision.String(), squashfsPath, whereDir, "squashfs"); err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply overlord/snapstate/snapmgr.go:1125]: I can extract this in a function. That can look better.

But in that extracted function, I do not see how it will be more obvious that sysd gets set when needed. Unless I missed something here.

File: overlord/snapstate/snapmgr.go [hunk 36]
```diff
@@ -1067,6 +1071,64 @@ func (m *SnapManager) localInstallCleanup() error {
...
+
+	// only run after we are seeded
+	var seeded bool
+	err := m.state.Get("seeded", &seeded)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:1099]: we have code like this in a few places, so probably they would need to switch to the helper. My slightly worry is that people might forget that 4 lines unexported helpers exist but I don't have a strong opinion. This sits exactly at the border of where a helper might or might not make sense.

File: overlord/snapstate/snapmgr.go (hunk 35 above)
[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/snapstate/snapmgr.go:1125]: In the extracted function you can short-circuit the logic 
```
if len(allStates) == 0 {
   return nil
}

if snapdenv.Preseeding() {
    sysd = systemd.NewEmulationMode(dirs.GlobalRootDir)
} else {
    sysd = systemd.New(systemd.SystemMode, nil)
}
```

Now it's easy to observe that sysd is set in all cases.

File: overlord/snapstate/snapmgr.go (hunk 36 above)
[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/snapstate/snapmgr.go:1099]: It's not important that the other places we do this switch now, it's about seperating the concerns of functions and making sure we have code that is easy to read and maintain.

[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:1099]: Sorry but we have lived without the helper for a while, so either the helper is an improvement that is worth applying consistently or is not. As I said, I don't have a strong opinion that we don't want it, it's marginal. But I do care about consistency. We do read things from state in many places and we don't introduce helpers for each. We considered having better Get functions to start with, but never got past the activation threshold to work out something for real in that direction.

File: overlord/snapstate/snapmgr.go (hunk 35 above)
[the remark this answers is not in the document]
Comment (valentindavid) [reply overlord/snapstate/snapmgr.go:1125]: Thank you.

File: overlord/snapstate/snapmgr.go (hunk 36 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:1099]: Had a longer discussion with @Meulengracht offline, it seems the real win would be more to later introduce a GetBool method parallel to Get and adopt it at in the most relevant places at least.

[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapmgr.go:1099]: In this regard, I would leave things as they are here.

[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/snapstate/snapmgr.go:1099]: Yep, I will resolve this for now

File: overlord/snapstate/snapmgr.go (hunk 35 above)
[the remark this answers is not in the document]
Comment (valentindavid) [reply overlord/snapstate/snapmgr.go]: I have changed the code so that `sysd` is never nil. But I only extracted the initialization of it.
